examples/qwen3_vl/qwen3_vl.zig

```diff
@@ -0,0 +1,1314 @@
+                .kernel_output_feature_dimension = 0,
+                .kernel_spatial_dimensions = &.{ 2, 3, 4 },
+                .output_batch_dimension = 0,
+                .output_feature_dimension = 1,
+                .output_spatial_dimensions = &.{ 2, 3, 4 },
+                .feature_group_count = 1,
+                .batch_group_count = 1,
+            },
+            loc,
+        );
+        if (self.bias) |b| y = y.add(b.convert(y.dtype()).broadcast(y._shape, &.{1}));
+        return y;
+    }
+};
+
+pub const VisionBlock = struct {
+    norm1: zml.nn.LayerNorm,
+    norm2: zml.nn.LayerNorm,
+    attn: VisionAttention,
+    mlp: VisionMlp,
+    num_heads: u32,
+
+    pub fn forward(self: VisionBlock, hidden_states: Tensor, cos: Tensor, sin: Tensor) Tensor {
+        const x = zml.call(self.norm1, .forward, .{hidden_states});
+        //Here we need to squeeze the output of the attention to remove the bs size because the image processing is not done in batch
+        //To be discussed when the model will handle several images and several sequences
+        const x1 = hidden_states.add(zml.call(self.attn, .forward, .{ x, cos, sin }).squeeze(0));
+        const x2 = zml.call(self.norm2, .forward, .{x1});
+        const x3 = x1.add(zml.call(self.mlp, .forward, .{x2}));
+
+        return x3.reuseBuffer(hidden_states);
+    }
+};
+
+// Vision patch embedding
+// Project the image into a visual embedding by applying a 3D convolution along the temporal and spatial dims
+pub const VisionPatchEmbed = struct {
+    proj: Conv3d,
+    patch_size: u32 = 14,
+    temporal_patch_size: u32 = 2,
+    in_channels: u32 = 3,
+    embed_dim: u32 = 1152,
+    pub fn init(
+        allocator: std.mem.Allocator,
+        patch_size: u32,
+        temporal_patch_size: u32,
+        in_channels: u32,
+        embed_dim: u32,
+        store: zml.aio.BufferStore,
+    ) !VisionPatchEmbed {
+        var conv3d = try zml.aio.populateModelWithPrefix(Conv3d, allocator, store, "model.visual.patch_embed.proj");
+        conv3d.temporal_stride = temporal_patch_size;
+        conv3d.spatial_stride = patch_size;
+
+        return .{
+            .proj = conv3d,
+            .patch_size = patch_size,
+            .temporal_patch_size = temporal_patch_size,
+            .in_channels = in_channels,
+            .embed_dim = embed_dim,
+        };
+    }
+
+    pub fn forward(self: VisionPatchEmbed, hidden_states: Tensor) Tensor {
+        const reshaped = hidden_states.reshape(.{ -1, self.in_channels, self.temporal_patch_size, self.patch_size, self.patch_size });
+        const conv_output = zml.call(self.proj, .forward, .{reshaped});
+        const reshaped_output = conv_output.reshape(.{ conv_output.dim(0), conv_output.dim(1) });
+
+        return reshaped_output;
+    }
+};
+
+pub const VisionRotaryEmbedding = struct {
+    rope_opts: zml.nn.RopeOpts,
+    dim: u32,
+
+    pub fn init(allocator: std.mem.Allocator, dim: u32, theta: f32) !VisionRotaryEmbedding {
```

Comment:
why allocatoor ? 

my feeling is that this struct should not exist, put the options directly into the parent struct,
and the forward can be inlined or maybe a helper function.